src/libraries/System.Data.Common/src/System/Data/Common/DbDataSource.cs

```diff
@@ -0,0 +1,570 @@
+                }
+                finally
+                {
+                    try
+                    {
+                        await _connection.CloseAsync().ConfigureAwait(false);
+                    }
+                    catch (Exception e)
+                    {
+                        ExceptionBuilder.TraceExceptionWithoutRethrow(e);
+
+                        // Swallow to allow the original exception to bubble up
+                        // Also, refrain from bubbling up the close exception even if there's no original exception,
+                        // since it's not relevant to the user - execution did complete successfully, and the connection
+                        // close is just an internal detail that shouldn't cause user code to fail.
+                    }
+                }
+            }
+
+            protected override DbDataReader ExecuteDbDataReader(CommandBehavior behavior)
+            {
+                _connection.Open();
+
+                try
+                {
+                    return _wrappedCommand.ExecuteReader(behavior | CommandBehavior.CloseConnection);
+                }
+                catch
+                {
+                    try
+                    {
+                        _connection.Close();
+                    }
+                    catch (Exception e)
+                    {
+                        ExceptionBuilder.TraceExceptionWithoutRethrow(e);
+
+                        // Swallow to allow the original exception to bubble up
+                    }
+
+                    throw;
+                }
+            }
+
+            protected override async Task<DbDataReader> ExecuteDbDataReaderAsync(
+                CommandBehavior behavior,
+                CancellationToken cancellationToken)
+            {
+                await _connection.OpenAsync(cancellationToken).ConfigureAwait(false);
+
+                try
+                {
+                    return await _wrappedCommand.ExecuteReaderAsync(
+                            behavior | CommandBehavior.CloseConnection,
+                            cancellationToken)
+                        .ConfigureAwait(false);
+                }
+                catch
+                {
+                    try
+                    {
+                        await _connection.CloseAsync().ConfigureAwait(false);
+                    }
+                    catch (Exception e)
+                    {
+                        ExceptionBuilder.TraceExceptionWithoutRethrow(e);
+
+                        // Swallow to allow the original exception to bubble up
+                    }
+
+                    throw;
+                }
+            }
+
+            protected override DbParameter CreateDbParameter()
+                => _wrappedCommand.CreateParameter();
+
+            public override void Cancel()
+                => _wrappedCommand.Cancel();
+
+            public override void Prepare()
+                => _wrappedCommand.Prepare();
```

Comment:
In some providers (e.g., MySqlConnector), `Prepare(Async)` can only be called if the command is associated with an open connection. (In MySQL Server, a prepared command is strongly associated with just one server session, so it needs to be known which session the command should be prepared on. Moreover, it doesn't make sense to open a connection, prepare the command, then close the connection without executing the command.)

So this method as written will fail with MySqlConnector. (And maybe other providers, too; I haven't checked them.)

Maybe that's MySqlConnector's problem to deal with, and it should either document that failure as a "known issue", or implement its own `DbCommandWrapper` that opens the connection in `Prepare`, keeps it open, and reuses that open connection in `ExecuteXyz`?

For now, I don't think the core logic in DbDataSource needs to be adjusted to accommodate this, but simply raising the issue for awareness, in case it is very common across most ADO.NET providers.

(The workaround I envision would be setting a `_isOpen` field to `true` in `Prepare`, then checking that field in other methods so `Open` isn't called twice. However, this is probably overkill since calling `Prepare` on a one-shot command is almost certainly pointless in most use cases.)

Reply:
Alternatively, `throw ExceptionBuilder.NotSupportedOnDataSourceCommand();` in this implementation, and let providers opt into connectionless Prepare (via a specialised class) if they can support it. This approach might make sense if most providers can't support this.

Reply:
@bgrainger yeah, when I originally thought about this, I thought the base implementation should just delegate and let the provider throw, which is what most providers do (SqlClient doesn't throw for Prepare without a connection, but that seems to just be a no-op, which isn't useful behavior either). But the provider exception ("no open connection") is likely to be a bit cryptic in this context, and ExceptionBuilder.NotSupportedOnDataSourceCommand seems more informative.

In the DbDataSource world, Prepare on a connectionless command could mean "prepare this on the server in general", for databases where prepared statements aren't bound to specific connections (FWIW PostgreSQL is the same as MySQL in this respect, and AFAIK most databases). So there's at least some reasonable implementation for this - though it would definitely require a specific data-source (connectionless) implementation, and simple delegation wouldn't work out-of-the-box.

Another way to look at this (as you hint), is to open a connection and keep it open until the command is disposed. This would allow multiple prepared executions via the same "connectionless" command without having to deal with a connection, but it goes against the connectionless model we're introduced here. Another issue here is that it's (unfortunately) fairly common for users (and even ORMs, specifically EF6) to not dispose commands; in this case, it would mean that the connection is never disposed, causing a leak. I'd probably avoid this just for that.

So I'll change the default implement to throw ExceptionBuilder.NotSupportedOnDataSourceCommand() as you suggest; providers with connectionless prepare can indeed override and implement their logic in any case.